greengrass_defender_agent/agent.py

```diff
@@ -103,16 +144,45 @@ def publish_metrics(ipc_client, config_changed, metrics_collector, sample_interv
         raise e
 
 
+def handle_expcetion_and_sleep(retry_action, exception, retry_time_sec, retry_count):
+    """
+    Handle exception and sleep for retry
+
+    :param retry_action: the name of retry action to print in the log
+    :param exception: the exception thrown from orginal action
+    :param retry_time_sec: number of seconds to sleep before the next retry
+    """
+    config.logger.exception(
+                "Exception occurred during the {} : {}".format(retry_action, exception)
+                )
+    config.logger.info(
+                "Will retry {} in {} seconds, retry count remaining: {}".format(retry_action, retry_time_sec, retry_count)
+            )
+    sleep(retry_time_sec)
+
+
 def main():
     # Get the ipc client
     ipc_client = ipc_utils.IPCUtils()
-    try:
-        ipc_client.connect()
-    except Exception as e:
-        config.logger.error(
-            "Exception occurred during the creation of an IPC client: {}".format(e)
-        )
-        exit(1)
+
+    # Connect to the GG Nucleus
+    
+    need_retry = True
+    retry_time = config.INITIAL_RETRY_INTERVAL_SECONDS
+    env_config = set_env_variables_config()
+    connect_retry = env_config[config.CONNECT_AND_PUBLISH_RETRY_CONFIG_KEY]
+
+    while (need_retry and connect_retry > 0):
+        connect_retry -= 1
+        try:
+            ipc_client.connect()
```

Comment:
this won't ever fail. If it fails, then Greengrass isn't running and you ought to exit. I don't think you need to wrap this with retries, unless the "connect" method is doing more than just connecting.

Reply:
It's only doing the connect(). I saw customer indicating that it would fail at IPC client creation (or connect? @BolongZhang-AWS can confirm what he did to reproduce) when they "deploy the Device Defender Component to a Greengrass Core Device and then reboot the device or restart the Greengrass systemd service while the device is not connected to the internet. "

Reply:
Performed testing with Michael yesterday afternoon. We actually found there's a TimeoutError thrown from this line with ipc client trying to connect. And after extend the IPC_CONNECT_TIMEOUT to 120 seconds the error disappeared. I think that's actually the root cause for the customer's issue.

Also I learned that when I retry the call to connect to ipc client, it is creating resources which are never closed. Discussed with Michael and his suggestion is to bound the retries.